core/flyout.js

```diff
@@ -209,44 +232,77 @@ Blockly.Flyout.prototype.getMetrics_ = function() {
     // Flyout is hidden.
     return null;
   }
-  var viewHeight = this.height_ - 2 * this.SCROLLBAR_PADDING;
-  var viewWidth = this.width_;
+
   try {
     var optionBox = this.workspace_.getCanvas().getBBox();
   } catch (e) {
     // Firefox has trouble with hidden elements (Bug 528969).
     var optionBox = {height: 0, y: 0};
   }
-  return {
+
+  var absoluteTop = this.verticalOffset_ + this.SCROLLBAR_PADDING
+  if (this.horizontalLayout_) {
+    if (this.toolboxPosition_ == Blockly.TOOLBOX_AT_BOTTOM) {
+      absoluteTop = 0;
+    }
+    var viewHeight = this.height_;
+    var viewWidth = this.width_ - 2 * this.SCROLLBAR_PADDING;
+  } else {
+    var viewHeight = this.height_ - 2 * this.SCROLLBAR_PADDING;
+    var viewWidth = this.width_;
+  }
+
+  var metrics = {
     viewHeight: viewHeight,
     viewWidth: viewWidth,
-    contentHeight: (optionBox.height + optionBox.y) * this.workspace_.scale,
+    contentHeight: (optionBox.height) * this.workspace_.scale,
+    contentWidth: (optionBox.width) * this.workspace_.scale,
     viewTop: -this.workspace_.scrollY,
-    contentTop: 0,
-    absoluteTop: this.SCROLLBAR_PADDING,
-    absoluteLeft: 0
+    viewLeft: -this.workspace_.scrollX,
+    contentTop: optionBox.y,
+    contentLeft: 0,
+    absoluteTop: absoluteTop,
+    absoluteLeft: this.SCROLLBAR_PADDING
   };
+  return metrics;
 };
 
 /**
- * Sets the Y translation of the flyout to match the scrollbars.
- * @param {!Object} yRatio Contains a y property which is a float
- *     between 0 and 1 specifying the degree of scrolling.
+ * Sets the translation of the flyout to match the scrollbars.
+ * @param {!Object} xyRatio Contains a y property which is a float
+ *     between 0 and 1 specifying the degree of scrolling and a
+ *     similar x property.
  * @private
  */
-Blockly.Flyout.prototype.setMetrics_ = function(yRatio) {
+Blockly.Flyout.prototype.setMetrics_ = function(xyRatio) {
   var metrics = this.getMetrics_();
   // This is a fix to an apparent race condition.
   if (!metrics) {
     return;
   }
-  if (goog.isNumber(yRatio.y)) {
+  if (!this.horizontalLayout_ && goog.isNumber(xyRatio.y)) {
     this.workspace_.scrollY =
-        -metrics.contentHeight * yRatio.y - metrics.contentTop;
+        -metrics.contentHeight * xyRatio.y - metrics.contentTop;
+  } else if (this.horizontalLayout_ && goog.isNumber(xyRatio.x)) {
+    if (this.RTL) {
```

Comment:
Suggest:
this.workspace_.scrollX = -metrics.contentWidth * xyRatio.x + (this.RTL ? metrics.contentLeft : -metrics.contentLeft);